File: Vuforia Spatial Toolbox/WebView/WebServerManager.m
//
//  WebServerManager.m
//  Vuforia Spatial Toolbox
//
//  Created by Benjamin Reynolds on 4/2/18.
//
// This is a singleton class that manages a NodeRunner instance which runs a local instance of the Spatial Edge Server

#import "WebServerManager.h"
#import "NodeRunner.h"

@implementation WebServerManager

#pragma mark - Singleton Methods

+ (id)sharedManager {
    static WebServerManager *sharedMyManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedMyManager = [[self alloc] init];
    });
    return sharedMyManager;
}

- (id)init {
    if (self = [super init]) {
        
        // Create server
        
        NSThread* nodejsThread = nil;
        nodejsThread = [[NSThread alloc]
                        initWithTarget:self
                        selector:@selector(startNode)
                        object:nil
                        ];
        // Set stack space for the Node.js thread, measured in bytes
        [nodejsThread setStackSize:8*1024*1024];
        [nodejsThread start];
        
    }
    return self;
}

- (void)startNode {
    NSLog(@"startNode");
    NSString* srcPath = [[NSBundle mainBundle] pathForResource:@"vuforia-spatial-edge-server/server.js" ofType:@""];
    NSLog(@"Path: %@", srcPath);
    NSArray* nodeArguments = [NSArray arrayWithObjects:
                              @"node",
                              srcPath,
                              nil
                              ];
    [NodeRunner startEngineWithArguments:nodeArguments];
    NSLog(@"[NodeRunner startEngineWithArguments:%@]", nodeArguments);
}

- (NSURL *)getServerURL {
     return [NSURL URLWithString:@"http://127.0.0.1:49368/"];
}

@end

File: Vuforia Spatial Toolbox/WebView/WebServerManager.swift
//
//  WebServerManager.swift
//  vst-swift
//
//  Created by Ben Reynolds on 9/8/21.
//

import Foundation

class WebServerManager {
    
    // MARK: - Singleton
    static let shared = WebServerManager()
    
    // MARK: - Initialization
    
    private init() {
        let nodejsThread = Thread(target: self, selector: #selector(self.startNode), object: nil)
        nodejsThread.stackSize = 8 * 1024 * 1024;
        nodejsThread.start()
        print("Init nodejsThread");
    }
    
    @objc private func startNode() {
        print("NodeRunner.startEngine");
        let srcPath = Bundle.main.path(forResource: "vuforia-spatial-edge-server/server.js", ofType: "")
        let nodeArguments:[String?] = ["node", srcPath]
        NodeRunner.startEngine(withArguments: nodeArguments as [Any])
    }
    
    // MARK: - Public Functions

    func getServerURL() -> URL {
        return URL(string: "http://127.0.0.1:49368/")!
    }
}
